File: Oven2Door-Mobile/lib/core/services/auth_service.dart
library oven2door_mobile.core.services.auth_service;

import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart';
import '../../api/auth_api.dart';

/// Result object for authentication attempts
class AuthResult {
  final bool success;
  final String message;

  AuthResult({required this.success, required this.message});
}

/// AuthService handles login/logout and token management.
class AuthService extends ChangeNotifier {
  AuthService({FirebaseAuth? firebaseAuth, AuthApi? authApi})
      : _auth = firebaseAuth ?? FirebaseAuth.instance,
        _authApi = authApi ?? AuthApi();

  final FirebaseAuth _auth;
  final AuthApi _authApi;

  User? get currentUser => _auth.currentUser;
  bool get isLoggedIn => currentUser != null;

  Future<AuthResult> signIn({
    required String email,
    required String password,
  }) async {
    try {
      await _auth.signInWithEmailAndPassword(email: email, password: password);
      notifyListeners();
      return AuthResult(success: true, message: 'Login successful');
    } on FirebaseAuthException catch (e) {
      return AuthResult(success: false, message: e.message ?? 'Login failed');
    } catch (e) {
      return AuthResult(success: false, message: 'Unexpected error: $e');
    }
  }

  Future<AuthResult> signUp({
    required String email,
    required String password,
    required String firstName,
    required String lastName,
    required String phone,
  }) async {
    try {
      final credential = await _auth.createUserWithEmailAndPassword(
        email: email,
        password: password,
      );
      final idToken = await credential.user?.getIdToken();
      if (idToken == null) {
        return AuthResult(
          success: false,
          message: 'Could not verify the new Firebase account.',
        );
      }

      await _authApi.signUp(
        idToken: idToken,
        firstName: firstName,
        lastName: lastName,
        email: email,
        phone: phone,
      );
      notifyListeners();
      return AuthResult(success: true, message: 'Account created successfully');
    } on FirebaseAuthException catch (e) {
      return AuthResult(success: false, message: e.message ?? 'Signup failed');
    } on AuthApiException catch (e) {
      return AuthResult(
        success: false,
        message: 'Firebase account was created, but MySQL sync failed: ${e.message}',
      );
    } catch (e) {
      return AuthResult(success: false, message: 'Unexpected error: $e');
    }
  }

  Future<void> signOut() async {
    await _auth.signOut();
    notifyListeners();
  }
}

File: src/app/authentication/services/auth.service.ts
import { Injectable } from '@angular/core';
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signOut,
  updateProfile,
  User
} from 'firebase/auth';
import { auth } from '../../firebase.config';
import { ApiService } from '../../services/api.service';

export interface AuthResult {
  success: boolean;
  message: string;
}

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  constructor(private api: ApiService) {}

  async signUp(
    email: string,
    password: string,
    firstName: string,
    lastName: string,
    phone: string
  ): Promise<AuthResult> {
    try {
      const credential = await createUserWithEmailAndPassword(auth, email, password);
      const user = credential.user;

      await updateProfile(user, {
        displayName: `${firstName} ${lastName}`
      });

      const idToken = await user.getIdToken();
      if (!idToken) {
        return {
          success: false,
          message: 'Could not verify the new Firebase account.'
        };
      }

      await this.api.signUp({
        idToken,
        firstName,
        lastName,
        email,
        phone
      });

      return { success: true, message: 'Account created successfully' };
    } catch (error: any) {
      if (error?.code === 'auth/email-already-in-use') {
        return { success: false, message: 'This email address is already registered.' };
      }
      if (error?.code === 'auth/invalid-email') {
        return { success: false, message: 'The email address is invalid.' };
      }
      if (error?.code === 'auth/weak-password') {
        return { success: false, message: 'The password is too weak.' };
      }
      if (error?.code === 'auth/network-request-failed') {
        return { success: false, message: 'Network error. Please check your internet connection.' };
      }
      return {
        success: false,
        message: error?.message || 'Signup failed'
      };
    }
  }

  async login(email: string, password: string): Promise<AuthResult> {
    try {
      await signInWithEmailAndPassword(auth, email, password);
      return { success: true, message: 'Login successful' };
    } catch (error: any) {
      return {
        success: false,
        message: error?.message || 'Login failed'
      };
    }
  }

  async logout(): Promise<void> {
    await signOut(auth);
  }

  getCurrentUser(): User | null {
    return auth.currentUser;
  }

  async getIdToken(): Promise<string | null> {
    const user = auth.currentUser;
    if (!user) {
      return null;
    }
    return user.getIdToken();
  }
}
